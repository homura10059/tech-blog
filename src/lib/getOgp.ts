import axios from 'axios'
import { JSDOM } from 'jsdom'

import { getAmazonImageUrl, getAmazonShortUrl } from './amazon'
import { readCache, writeCache } from './cache'

export type OgpMeta = {
  title: string
  description: string
  image: string
  url: string
}

const trimTitle = (title: string) => title.trim()

const getImageUrl = (imageUrl: string, url: string) => {
  if (imageUrl.charAt(0) !== '/') return imageUrl
  const parsed = new URL(url)
  return `https//${parsed.hostname}${imageUrl}`
}

type HasImageURL = {
  'og:image:secure_url'?: string
  'og:image'?: string
}
const getOgpImageUrl = (mergedData: HasImageURL, encodedUri: string) => {
  const imageUrl =
    mergedData['og:image:secure_url'] ?? mergedData['og:image'] ?? ''
  return getImageUrl(imageUrl, encodedUri)
}

export const getOgp = async (url: string): Promise<OgpMeta> => {
  const isAmazon = url.includes('https://www.amazon.co.jp/')
  const encodedUri = encodeURI(isAmazon ? getAmazonShortUrl(url) ?? url : url)

  const cache = await readCache(encodedUri)
  if (cache) return cache

  const headers = { 'User-Agent': 'bot' }

  const res = await axios.get(encodedUri, { headers: headers })
  const html = res.data
  const dom = new JSDOM(html)
  const meta = isAmazon
    ? dom.window.document.body.querySelectorAll('meta')
    : dom.window.document.head.querySelectorAll('meta')

  const metaTags = Array.from(meta.values())
  const metaData = Object.fromEntries(
    metaTags
      .filter(element => element.name !== '')
      .map(element => [element.name, element.content])
  )
  const propertyData = Object.fromEntries(
    metaTags
      .filter(element => element.hasAttribute('property'))
      .map(element => [
        element.getAttribute('property')?.trim(),
        element.content
      ])
  )

  const mergedData = { ...metaData, ...propertyData }

  const ogpMeta: OgpMeta = {
    title: trimTitle(mergedData['og:title'] ?? mergedData.title),
    description: mergedData['og:description'] ?? mergedData.description,
    image: isAmazon
      ? getAmazonImageUrl(encodedUri) ?? ''
      : getOgpImageUrl(mergedData, encodedUri),
    url: encodedUri
  }

  await writeCache(ogpMeta)

  return ogpMeta
}
