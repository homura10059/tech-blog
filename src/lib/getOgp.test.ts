import { expect, test } from 'vitest'

import { getOgp } from './getOgp'

test.each`
  url
  ${'https://tech-blog.homura10059.dev/'}
  ${'https://tech-blog.homura10059.dev/posts/2022-04-23-1'}
  ${'https://www.amazon.co.jp/dp/B09RQGMYKZ/'}
  ${'https://www.bbiq.jp/'}
  ${'https://preducts.jp/products/mount-switch'}
`('getOgp for $url', async ({ url }) => {
  const actual = await getOgp(url)
  expect(actual).toMatchSnapshot()
})
