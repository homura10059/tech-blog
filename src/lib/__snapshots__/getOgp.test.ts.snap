// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getOgp for 'https://preducts.jp/products/mount-sw…' 1`] = `
{
  "description": "PREDUCTS DESKにNintendo Switchを装着するための専用マウント ※レールに装着するためのBase Sliderは付属しません",
  "image": "http://res.cloudinary.com/preducts-inc/image/upload/w_1200/w_1920,f_auto/shop/photo/items/mount-switch/mount-switch-ogp.jpg",
  "title": "Mount for Switch",
  "url": "https://preducts.jp/products/mount-switch",
}
`;

exports[`getOgp for 'https://tech-blog.homura10059.dev/' 1`] = `
{
  "description": "programming memo & tech posts.",
  "image": "https://i.imgur.com/BqDJIrtm.png",
  "title": "homura's tech blog",
  "url": "https://tech-blog.homura10059.dev/",
}
`;

exports[`getOgp for 'https://tech-blog.homura10059.dev/pos…' 1`] = `
{
  "description": "このブログ内のリンクを Next.js の <Link> や Twitter の Embed に変換したいので対応した",
  "image": "https://i.imgur.com/W7bKLLUm.jpeg",
  "title": "ブログ内のリンクを Next.js の <Link> や Embed に変換したい",
  "url": "https://tech-blog.homura10059.dev/posts/2022-04-23-1",
}
`;

exports[`getOgp for 'https://www.amazon.co.jp/dp/B09RQGMYK…' 1`] = `
{
  "description": "Amazon.co.jp: コンセプトから理解するRust eBook : 原 旅人: Kindleストア",
  "image": "https://images-na.ssl-images-amazon.com/images/P/B09RQGMYKZ.09.LZZZZZZZ.jpg",
  "title": "Amazon.co.jp: コンセプトから理解するRust eBook : 原 旅人: Kindleストア",
  "url": "https://www.amazon.co.jp/dp/B09RQGMYKZ?tag=homura10059-22",
}
`;

exports[`getOgp for 'https://www.bbiq.jp/' 1`] = `
{
  "description": "プロバイダー＆光インターネット回線の「月額料金」がずっとおトクな九州のプロバイダーサービスBBIQ(ビビック)のご案内。BBIQなら、光回線とプロバイダー料金、全部含んで3000円台！セキュリティソフトや開通時の訪問設定サービスも無料です！BBIQ公式サイト。",
  "image": "https//www.bbiq.jp/common/img/og.jpg",
  "title": "BBIQ(ビビック)公式サイト | 九州の光インターネット回線＆プロバイダー",
  "url": "https://www.bbiq.jp/",
}
`;
